src/crimson/osd/replicated_recovery_backend.cc

```diff
@@ -27,7 +27,24 @@ seastar::future<> ReplicatedRecoveryBackend::recover_object(
   assert(is_recovering(soid));
   // start tracking the recovery of soid
   return maybe_pull_missing_obj(soid, need).then([this, soid, need] {
-    return maybe_push_shards(soid, need);
+    auto& recovery_waiter = recovering.at(soid);
+    if (recovery_waiter.obc) {
```

Comment:
this condition check is problematic, I've just submitted a [PR](https://github.com/ceph/ceph/pull/38646) for this, please take a look:-)